src/rdkafka_broker.c

```diff
@@ -3631,13 +3633,62 @@ static void rd_kafka_toppar_fetch_backoff (rd_kafka_broker_t *rkb,
                 return;
 
         rktp->rktp_ts_fetch_backoff = rd_clock() + (backoff_ms * 1000);
+
         rd_rkb_dbg(rkb, FETCH, "BACKOFF",
-                   "%s [%"PRId32"]: Fetch backoff for %dms: %s",
+                   "%s [%"PRId32"]: Fetch backoff for %dms%s %s",
                    rktp->rktp_rkt->rkt_topic->str, rktp->rktp_partition,
-                   backoff_ms, rd_kafka_err2str(err));
+                   backoff_ms,
+                   err ? ":" : "", 
+                   err ? rd_kafka_err2str(err) : "");
 }
 
 
+/**
+ * @brief Handle preferred replica in fetch response.
+ */
+void
+rd_kafka_fetch_preferred_replica_handle (rd_kafka_toppar_t *rktp,
+                                         rd_kafka_buf_t *rkbuf,
+                                         rd_kafka_broker_t *rkb,
+                                         int32_t preferred_id) {
+        if (rd_interval_immediate(&rktp->rktp_new_lease_intvl,
+                                  60*1000000/*1 minute*/, 0) < 0) {
+                rd_rkb_log(rkb, LOG_WARNING, "FETCH",
+                        "%.*s [%"PRId32"]: Not migrating to preferred "
+                        "replica %"PRId32" because previous lease was "
+                        "granted < 1 minute ago",
+                        RD_KAFKAP_STR_PR(rktp->rktp_rkt->rkt_topic),
+                        rktp->rktp_partition, preferred_id);
+
+                rd_kafka_toppar_fetch_backoff(rkb,
```

Comment:
this is partly addressing your concern above (logs being flooded with messages), though it's doing a fairly poor job of that. what are you thinking - a suppression `rd_interval` just for this? 

if we get here, something is up. we've bounced back from a preferred replica early indicating a problem there (though note: see below) and I think it's likely the leader is just going to bounce us right back again. regardless, this is an infrequent error scenario and 500ms is not a big deal in that context, so I was thinking better to err on the side of conservative. feel free to argue that's not a good way of looking at it.

I realized I made a mistake related to the above - this scenario can happen on a leader change since that could happen at any time after the preferred replica lease is granted (possibly before 1 minute). when that happens, there should be an `rd_interval_expedite` to prevent this case getting hit. will make that change.

Reply:
I've changed the behavior + added a comment.

I doubt you will like this but I don't know what's better. Note: there could be an unnecessary back-off in the event there was a problem on a preferred replica, the toppar was delegated back to the leader and the leader then delegates it to a different preferred replica, but I don't think the complexity in tracking that case is worth the benefit.

Reply:
I've added an additional dbg message instead - not sure what error code to use. the root cause is `RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE`, but that seems likely to cause confusion if used here.